pythran/analyses/range_values.py

```diff
@@ -0,0 +1,327 @@
+
+import ast
+import copy
+
+from pythran.analyses import Globals, Aliases
+from pythran.intrinsic import Intrinsic
+from pythran.passmanager import FunctionAnalysis
+from pythran.range import Range, UNKNOWN_RANGE, combine
+
+
+class RangeValues(FunctionAnalysis):
+
+    """
+    This analyse extract positive subscripts from code.
+
+    It is flow insensitif and aliasing is not taken into account as integer
+    doesn't create aliasing in Python.
+    """
+
+    def __init__(self):
+        """Initialize instance variable and gather globals name information."""
+        self.result = dict()
+        super(RangeValues, self).__init__(Globals, Aliases)
+
+    def add(self, variable, range_):
+        """
+        Add a new low and high bound for a variable.
+
+        As it is flow insensitif, it compares it with olds values and update it
+        if needed.
+        """
+        if variable not in self.result:
+            self.result[variable] = range_
+        else:
+            self.result[variable].update(range_)
+
+    def visit_FunctionDef(self, node):
+        """ Set default range value for globals and attributs.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(a, b): pass")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=inf)
+        """
+        for global_name in self.globals:
+            self.result[global_name] = UNKNOWN_RANGE
+        for attr in node.args.args:
+            self.result[attr.id] = UNKNOWN_RANGE
+
+        map(self.visit, node.body)
+
+    def visit_Assign(self, node):
+        """
+        Set range value for assigned variable.
+
+        We do not handle container values.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(): a = b = 2")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=2, high=2)
+        >>> res['b']
+        Range(low=2, high=2)
+        """
+        assigned_range = self.visit(node.value)
+        for target in node.targets:
+            if isinstance(target, ast.Name):
+                self.add(target.id, copy.deepcopy(assigned_range))
+
+    def visit_AugAssign(self, node):
+        """ Update range value for augassigned variables.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(): a = 2; a -= 1")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=1, high=2)
+        """
+        if isinstance(node.target, ast.Name):
+            self.result[node.target.id].update(
+                combine(self.result[node.target.id],
+                        self.visit(node.value),
+                        node.op))
```

Comment:
ok, because of this kind of in-place modification.
BTW, it seems to me you're not using alias information here, which could lead to inaccurate (false?) informations.